Add `runtime-wasm` pipeline

radical opened the pull request:
This adds support for a `runtime-wasm` pipeline, which uses
`runtime-extra-platforms.yml`.

The wasm jobs have been moved to `runtime-extra-platforms-wasm.yml`, and
the rest to `runtime-extra-platforms-other.yml`. And the two files are
included conditionally.

Effectively, this still runs all these jobs when
`runtime-extra-platforms` is run (same as before). But it allows running
just the wasm jobs, triggered manually with:

    `/azp run runtime-wasm`

When `runtime-wasm` is manually run:

- LibraryTests will run all the tests on linux+windows
- LibraryTests_EAT will run all the tests on linux
- LibraryTests_AOT will run all the tests on linux+windows
- Wasm.Build.Tests, and Wasm debugger tests will run on linux+windows
- Runtime tests will run on linux

When `runtime-extra-platforms` is manually run:

- LibraryTests will run all the tests, if paths changed
- LibraryTests_EAT will run all the tests on linux, if paths changed
- LibraryTests_AOT will run all the tests on linux+windows, if paths changed
- Wasm.Build.Tests, and Wasm debugger tests will never run
- Runtime tests will never run

On rolling builds, `runtime-extra-platforms` runs by default:

- LibraryTests will run all the tests
- LibraryTests_EAT will run all the tests on linux
- LibraryTests_AOT will run all the tests on linux+windows
- Wasm.Build.Tests, and Wasm debugger tests will never run
- Runtime tests will never run

With feedback from @steveisok, and @safern.

radical committed:
Add `runtime-wasm` pipeline

This adds support for a `runtime-wasm` pipeline, which uses
`runtime-extra-platforms.yml`.

The wasm jobs have been moved to `runtime-extra-platforms-wasm.yml`, and
the rest to `runtime-extra-platforms-other.yml`. And the two files are
included conditionally.

Effectively, this still runs all these jobs when
`runtime-extra-platforms` is run (same as before). But it allows running
just the wasm jobs, triggered manually with:

    `/azp run runtime-wasm`

When `runtime-wasm` is manually run:

- LibraryTests will run all the tests on linux+windows
- LibraryTests_EAT will run all the tests on linux
- LibraryTests_AOT will run all the tests on linux+windows
- Wasm.Build.Tests, and Wasm debugger tests will run on linux+windows
- Runtime tests will run on linux

When `runtime-extra-platforms` is manually run:

- LibraryTests will run all the tests, if paths changed
- LibraryTests_EAT will run all the tests on linux, if paths changed
- LibraryTests_AOT will run all the tests on linux+windows, if paths changed
- Wasm.Build.Tests, and Wasm debugger tests will never run
- Runtime tests will never run

On rolling builds, `runtime-extra-platforms` runs by default:

- LibraryTests will run all the tests
- LibraryTests_EAT will run all the tests on linux
- LibraryTests_AOT will run all the tests on linux+windows
- Wasm.Build.Tests, and Wasm debugger tests will never run
- Runtime tests will never run

With feedback from @steveisok, and @safern.

radical commented:
This is the new state:

* jobs with `only-pc` are run only if the path checks are true

# PR:
| .                 | runtime               | runtime-staging         | runtime-extra-platforms(manual only) | wasm-only (manual only) |
| ----------------- | -------------         | ---------------         | ------------------------------------ | -------                 |
| libtests          | linux: all,   only-pc | windows: all,   only-pc | linux+windows: all, only-pc          | linux+windows: all      |
| libtests eat      | linux: smoke, only-pc | -                       | linux:         all, only-pc          | linux:         all      |
| libtests aot      | linux: smoke, only-pc | windows: smoke, only-pc | linux+windows: all, only-pc          | linux+windows: all      |
|                   |                       |                         |                                      |                         |
| Wasm.Build.Tests  | linux:        only-pc | windows:        only-pc | -                                    | linux+windows           |
| Debugger tests    | -                     | linux+windows:  only-pc | -                                    | linux+windows           |
| Runtime tests     | linux:        only-pc | -                       | -                                    | linux                   |

- none of the jobs are ever run unless the path checks are true

# Rolling build:

* jobs w/o `only-pc` are always run

| .                 | runtime            | runtime-staging               | runtime-extra-platforms (always run) | wasm-only (manual only) |
| ----------------- | -------------      | ---------------               | ------------------------------------ | --                      |
| libtests          | linux: all         | windows: all,         only-pc | linux+windows: all                   | NA                      |
| libtests eat      | linux: smoke(=all) | -                             | linux:         all                   |                         |
| libtests aot      | linux: smoke(=all) | windows: smoke(=all), only-pc | linux+windows: all                   |                         |
|                   |                    |                               |                                      |                         |
| Wasm.Build.Tests  | linux: all         | windows:              only-pc | linux+windows                        |                         |
| Debugger tests    | -                  | linux+windows                 | -                                    |                         |
| Runtime tests     | linux: all         | -                             | -                                    |                         |


There is an existing inconsistency in rolling builds:

- runtime is always running the tests
- runtime-staging is running the tests only on on path changes
- runtime-extra-platforms

radical commented:
/azp run wasm-only

radical commented:
/azp run wasm-only

radical commented:
/azp run wasm-only

radical commented:
/azp run wasm-only

radical commented:
/azp run wasm-only

radical commented:
/azp run wasm-only

radical commented:
/azp run wasm-only

radical commented:
/azp run wasm-only

radical commented:
/azp run wasm-only

radical commented:
/azp run wasm-only

safern commented:
May I suggest that we name it `runtime-wasm` to match all our other naming conventions?

jkotas commented on eng/pipelines/runtime-extra-platforms-non-wasm.yml:
```diff
@@ -0,0 +1,364 @@
+jobs:
```
I expect that people will want to create more subsets from this, for similar reasons you are creating the wasm subset.

Would it make sense to call this file `runtime-other` so that we do not end up with odd names as more subsets are created?

radical commented:
/azp run wasm-only

safern commented on eng/pipelines/common/xplat-setup.yml:
```diff
@@ -49,8 +49,8 @@ jobs:
       - name: _runSmokeTestsOnlyArg
-        value: '/p:RunSmokeTestsOnly=$(isNotExtraPlatformsBuild)'
-      
+        value: '/p:RunSmokeTestsOnly=$(isNotExtraPlatformsOrWasmOnlyBuild)'
```
Can we rename this variable to `RunSmokeTestsOnly` instead?

safern commented on eng/pipelines/runtime-extra-platforms-wasm-only.yml:
```diff
@@ -0,0 +1,225 @@
+jobs:
```
We could share the same yml file and just condition with `${{ }}` based on definition name. 

Look at: https://github.com/dotnet/runtime/blob/59e643b9932122a12e71ceb360a9ae3bb9420dec/eng/pipelines/libraries/outerloop.yml#L24

radical replied:
hm I'll have to leave that one to you

safern replied:
Never mind, you are already doing this but with an extra layer (using template references).

radical commented:
/azp run runtime-extra-platforms

radical commented:
/azp run runtime-wasm

radical commented:
/azp run runtime-extra-platforms

radical commented:
The next question, which can be resolved in a follow up PR too. Given:

# Rolling build:

* jobs w/o `only-pc` are always run

| .                 | runtime     | runtime-staging       | runtime-extra-platforms |
| ----------------- | ----------- | ---------------       | ----------------------- |
| libtests          | linux: all  | windows: all, only-pc | linux+windows: all      |
| libtests eat      | linux: all  | -                     | linux:         all      |
| libtests aot      | linux: all  | windows: all, only-pc | linux+windows: all      |
|                   |             |                       |                         |
| Wasm.Build.Tests  | linux: all  | windows:      only-pc | linux+windows           |
| Debugger tests    | -           | linux+windows         | -                       |
| Runtime tests     | linux: all  | -                     | -                       |

`runtime-extra-platforms` runs all the wasm jobs, unconditionally.
`runtime` is running some of them, unconditionally.
`runtime-staging` is running the rest, with path conditions.

I think we should run the ones on `runtime-staging`, unconditionally too. Then we can have all the wasm jobs in `runtime-extra-platforms`(when running as part of a rolling build) not run at all.

safern commented on eng/pipelines/runtime-extra-platforms.yml:
```diff
@@ -34,482 +34,9 @@ jobs:
-          eq(dependencies.evaluate_paths.outputs['SetPathVars_runtimetests.containsChange'], true),
-          eq(variables['isRollingBuild'], true))
+# Any jobs that are not specific to any platform
+- ${{ if or(eq(variables.isNotSpecificPlatformOnlyBuild, true), eq(variables.isRollingBuild, true)) }}:
```
is the `isRollingBuild` needed? I expect the `isNotSpecificPlatformOnlyBuild` to be true when we are in a rolling build.

radical committed:
Flow `isExtraPlatformsBuild` to the new *-wasm, and *-other templates

```
/mnt/vss/_work/1/s/.packages/microsoft.dotnet.helix.sdk/7.0.0-beta.22080.1/tools/Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(18,5): error : You must specify at least one target queue to send a job to helix. Use the HelixTargetQueues property or HelixTargetQueue items. [/mnt/vss/_work/1/s/src/libraries/sendtohelixhelp.proj]
```

Fix suggested by @safern

radical committed:
Update condition, based on feedback

radical committed:
Merge remote-tracking branch 'origin/main' into wasm-manual

radical commented:
/azp run runtime-wasm,runtime-extra-platforms

radical committed:
Fix typo

radical commented:
/azp run runtime-wasm,runtime-extra-platforms

radical commented:
- `Build Android arm Release AllSubsets_Mono` - `System.Net.Ping.Functional.Tests` failing due to `insufficient storage`
- `Build Android arm64 Release AllSubsets_Mono` - `Failed to pull file. Device is running Android API 30, trying fallback to pull /data/user/0/net.dot.System.Net.Ping.Functional.Tests/cache/testResults.xml`
- `Build Android x64 Release AllSubsets_Mono_RuntimeTests` - `fail: Non-success instrumentation exit code: 101, expected: 0`

radical committed:
Rename the var `RunSmokeTestsOnly` to *not* the match msbuild property

.. used to control smoke tests. The variables in yml seem to get
converted to environment variables which can affect the build. Like
causing unrelated builds to get `RUNSMOKETESTSONLY=true`, and breaking
the build.

radical committed:
Merge remote-tracking branch 'origin/main' into wasm-manual

radical commented:
/azp run runtime-wasm,runtime-extra-platforms

radical committed:
Merge remote-tracking branch 'origin/main' into wasm-manual

radical committed:
runtime-wasm: don't trigger for scheduled builds

radical commented:
/azp run runtime-wasm,runtime-extra-platforms

radical committed:
fix yml

safern commented on eng/pipelines/runtime-extra-platforms.yml:
```diff
@@ -13,7 +13,7 @@ variables:
   - template: /eng/pipelines/common/variables.yml
 
 - ${{ if eq(variables.isNotSpecificPlatformOnlyBuild, true) }}:
```
I don't think expand time conditions work for schedules. We would have to go to the AzDo UI for this pipeline and override the CI triggers/schedule.

steveisok replied:
I don't think it'll work either.  

If you want such a behavior and not use AzDo UI, I think you'd have to make a template and then have two pipelines.  One with a schedule and one w/o.

radical replied:
I disabled it from the UI.

radical committed:
Revert "fix yml"

This reverts commit 0aac4058969b0129bdf256572677f3b09f9191ff.

radical committed:
Revert "runtime-wasm: don't trigger for scheduled builds"

This reverts commit c929f5c6a8edf54942c61d39f9f1fea7f6fb3b37.

radical committed:
Merge remote-tracking branch 'origin/main' into wasm-manual

radical commented:
Merging main to pick up some test fixes for Android.
The windows/x64 test failures are failing on rolling builds too.

radical commented:
/azp run runtime-extra-platforms

radical commented:
Anything else needed for this PR?

premun commented:
@radical love the table overview!

radical commented:
> Does it need to be reflected in somewhere in docs, e.g. https://github.com/dotnet/runtime/blob/main/docs/pr-builds.md ?

I can add it in a follow up PR.
